bgpv1: Add `bgp/routes` API endpoint and `cilium bgp routes` CLI command

rastislavs opened the pull request:
Adds a new `bgp/routes` API endpoint and the `cilium bgp routes` CLI command implementation that can be used for retrieval of available/advertised routes from BGP Control Plane's RIBs.

**Example usage:**

Available `ipv4 unicast` routes:
```
$ cilium bgp routes available ipv4 unicast
VRouter   Prefix              NextHop   Age   Attrs
65001     10.0.0.0/24         0.0.0.0   21s   [{Origin: i} {Nexthop: 0.0.0.0}]   
65001     192.168.10.176/32   0.0.0.0   21s   [{Origin: i} {Nexthop: 0.0.0.0}]   
```

`ipv4 unicast` routes advertised to the peer `172.0.0.1`:
```
$ cilium bgp routes advertised ipv4 unicast peer 172.0.0.1
VRouter   Prefix              NextHop     Age      Attrs
65001     10.0.0.0/24         172.0.0.2   20m33s   [{Origin: i} {AsPath: 65001} {Nexthop: 172.0.0.2}]   
65001     192.168.10.176/32   172.0.0.2   20m33s   [{Origin: i} {AsPath: 65001} {Nexthop: 172.0.0.2}]
```

Available ipv4 unicast routes in virtual router with ASN `65001`:
```
$ cilium bgp routes available ipv4 unicast vrouter 65001
VRouter   Prefix              NextHop   Age     Attrs
65001     10.0.0.0/24         0.0.0.0   24m7s   [{Origin: i} {Nexthop: 0.0.0.0}]   
65001     192.168.10.176/32   0.0.0.0   24m7s   [{Origin: i} {Nexthop: 0.0.0.0}]   
```

Help:
```
$ cilium bgp routes -h
List routes in the BGP Control Plane's Routing Information Bases (RIBs)

Usage:
  cilium bgp routes <available | advertised> <afi> <safi> [vrouter <asn>] [peer|neighbor <address>] [flags]

Examples:
  Get all IPv4 unicast routes available:
    cilium bgp routes available ipv4 unicast

  Get all IPv6 unicast routes available for a specific vrouter:
    cilium bgp routes available ipv6 unicast vrouter 65001

  Get IPv4 unicast routes advertised to specific peer:
    cilium bgp routes advertised ipv4 unicast peer 10.0.0.1

Flags:
  -h, --help            help for routes
  -o, --output string   json| yaml| jsonpath='{}'

Global Flags:
      --config string   Config file (default is $HOME/.cilium.yaml)
  -D, --debug           Enable debug messages
  -H, --host string     URI to server-side API
```

YutaroHayakawa committed:
bgpv1: Add a new API endpoint bgp/routes

Add a new API endpoint bgp/routes to retrieve BGP CPlane's RIB from API.
The actual implementation will be in subsequent commits.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>
Co-authored-by: Rastislav Szabo <rastislav.szabo@isovalent.com>

YutaroHayakawa committed:
api,bgpv1: Implement bgp/routes API

Implement bgp/routes API by wiring-up API call with BGPRouterManager.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>
Co-authored-by: Rastislav Szabo <rastislav.szabo@isovalent.com>

YutaroHayakawa committed:
bgpv1: Implement `cilium bgp routes` command

Implement `cilium bgp routes` command that retrieves routes from
bgp/routes API and displays response.
Also allow using "peer" and "neighbor" keyword interchangeably in bgp CLI
(e.g. "cilium bgp peers" == "cilium bgp neighbors",
"cilium bgp routes ... peer 172.0.0.1" == "cilium bgp routes ... neighbor 172.0.0.1").

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>
Co-authored-by: Rastislav Szabo <rastislav.szabo@isovalent.com>

zacharysarah commented on Documentation/configuration/api-restrictions-table.rst:
```diff
@@ -29,6 +29,8 @@ DeleteServiceID       -
                       Cilium BGP virtual router. This includes session state,
                       uptime, information per address family, etc.
+GetBGPRoutes          Retrieves routes from BGP Control Plane RIB filtered by
+                      specified parameters
```
Use active voice.
```suggestion
GetBGPRoutes          Retrieves routes from BGP Control Plane RIB filtered by
                     parameters you specify
```

rastislavs replied:
done

zacharysarah reviewed (changes requested):
@rastislavs One small wording nit, otherwise LGTM for docs

dylandreimerink reviewed (approved):
Changes look good to me. It might be useful to add this command to the [bugtool output](https://github.com/cilium/cilium/blob/main/bugtool/cmd/configuration.go#L373) so the info is available in sysdumps.

rastislavs committed:
bgpv1: Gracefully handle bgp peers/routes API/CLI if BGP CP is disabled

Fixes an issue with bgp/peers & bgp/routes API endpoint and cilium
"bgp peers" & "bgp routes" CLI command when BGP Control Plane is disabled.
Instead of returning an internal server error returns an specific error
that can be properly handled in the CLI command code.

Signed-off-by: Rastislav Szabo <rastislav.szabo@isovalent.com>

rastislavs committed:
bugtool: Add "cilium bgp peers" & "cilium bgp routes" command outputs

Includes "cilium bgp peers" & "cilium bgp routes" command outputs
in the bugtool so that BGP info is available in sysdumps.

Signed-off-by: Rastislav Szabo <rastislav.szabo@isovalent.com>

zacharysarah reviewed (approved):
LGTM for docs

harsimran-pabla commented on cilium/cmd/bgp_route_get.go:
```diff
@@ -0,0 +1,217 @@
+			fmt.Fprintf(w, "%s\t", path.NLRI)
+			fmt.Fprintf(w, "%s\t", nextHopFromPathAttributes(path.PathAttributes))
+			fmt.Fprintf(w, "%s\t", time.Duration(path.AgeNanoseconds).Round(time.Second))
+			fmt.Fprintf(w, "%s\t", path.PathAttributes)
```
nit: We should probably limit how long this attr field can get. For complete attributes, users could just json output. For stdout, we should be careful as this can be very long and will make output pretty difficult to read.

rastislavs replied:
I have included that in the follow-up issue (https://github.com/cilium/cilium/issues/27420). As the PathAttributes are base64-encoded in the yaml/json output, there is no other way to show them in human-readable format ATM.

For now, IMO it is still the better option to overflow a line potentially, than not be able to show them. Anyway, for standard Cilium use-cases (service / PodCIDR announcements etc.) this is very unlikely to be longer than one line.

harsimran-pabla commented on pkg/bgpv1/manager/manager.go:
```diff
@@ -364,6 +366,55 @@ func (m *BGPRouterManager) GetPeers(ctx context.Context) ([]*models.BgpPeer, err
+	defer m.RUnlock()
+
+	// validate router ASN
+	if params.RouterAsn != nil {
```
m.Servers is a map, we should be able to do direct lookup for ASN.

harsimran-pabla commented on pkg/bgpv1/manager/manager_test.go:
```diff
@@ -0,0 +1,228 @@
+			require.Equal(t, len(routes), len(tt.expectedPrefixes), "retrieved routes count does not match expected prefix count")
+
+			// ensure we see expected prefixes in retrieved routes
+			for _, cidr := range tt.expectedPrefixes {
```
can we simplify this by require.Contains

rastislavs replied:
actually simplified even more with `EqualValues`

harsimran-pabla reviewed (commented):
Few comments, overall looks good. There are few follow ups, maybe tasks can be created for them
- Extend this to cilium-cli
- Paging of the result, as it can be very large (order of 10k).
- Some optimization for later ( getting result concurrently from various vrouters - gobgp instances )

harsimran-pabla reviewed (approved):
Thank you!! LGTM

nathanjsweet commented on api/v1/openapi.yaml:
```diff
@@ -1079,6 +1079,43 @@ paths:
+  "/bgp/routes":
+    get:
+      summary: Lists BGP routes from BGP Control Plane RIB.
+      description: Retrieves routes from BGP Control Plane RIB filtered by parameters you specify
```
nit: Punctuation; no need for first person pronoun. 
```suggestion
      description: Retrieves routes from BGP Control Plane RIB, filtered by specified parameters.
```

rastislavs replied:
Hah funny, that is what it was originally, but was reworded by @zacharysarah as part of docs review: https://github.com/cilium/cilium/pull/27182#discussion_r1281254748

I guess the challenge here is that docs are auto-generated from the API description, and the wording that works for docs does not work that well for API :)

nathanjsweet replied:
Funny! Okay. Pay attention to her over me. She's our docs expert.

nathanjsweet reviewed (changes requested):
A couple of nits for API and CLI.

nathanjsweet reviewed (approved):
Approve for API and CLI
